File: Progs/PhD/src/ILP/examples/src/cs/QPex2.cs
// --------------------------------------------------------------------------
// File: QPex2.cs
// Version 12.2
// --------------------------------------------------------------------------
// Licensed Materials - Property of IBM
// 5725-A06 5725-A29 5724-Y48 5724-Y49 5724-Y54 5724-Y55
// Copyright IBM Corporation 2003, 2010. All Rights Reserved.
//
// US Government Users Restricted Rights - Use, duplication or
// disclosure restricted by GSA ADP Schedule Contract with
// IBM Corp.
// --------------------------------------------------------------------------
//
// QPex2.cs - Reading in and optimizing a QP problem
//
// To run this example, command line arguments are required.
// i.e.,   QPex2  filename method
// where
//     filename is the name of the file, with .mps, .lp, or .sav extension
//     method   is the optimization method
//                 o          default
//                 p          primal simplex
//                 d          dual   simplex
//                 b          barrier without crossover
//                 n          network with dual simplex cleanup
//
// Example:
//     QPex2  qpex.lp o
//

using ILOG.Concert;
using ILOG.CPLEX;
using System.Collections;


public class QPex2 {
   internal static void Usage() {
      System.Console.WriteLine("usage:  QPex2 <filename> <method>");
      System.Console.WriteLine("          o       default");
      System.Console.WriteLine("          p       primal simplex");
      System.Console.WriteLine("          d       dual   simplex");
      System.Console.WriteLine("          b       barrier without crossover");
      System.Console.WriteLine("          n       network with dual simplex cleanup");
   }

   public static void Main(string[] args) {
      if ( args.Length != 2 ) {
         Usage();
         return;
      }
      try {
         Cplex cplex = new Cplex();

         // Evaluate command line option and set optimization method accordingly.
         switch ( args[1].ToCharArray()[0] ) {
         case 'o': cplex.SetParam(Cplex.IntParam.RootAlg,
                                  Cplex.Algorithm.Auto);
                   break;
         case 'p': cplex.SetParam(Cplex.IntParam.RootAlg,
                                  Cplex.Algorithm.Primal);
                   break;
         case 'd': cplex.SetParam(Cplex.IntParam.RootAlg,
                                  Cplex.Algorithm.Dual);
                   break;
         case 'b': cplex.SetParam(Cplex.IntParam.RootAlg,
                                  Cplex.Algorithm.Barrier);
                   cplex.SetParam(Cplex.IntParam.BarCrossAlg,
                                  Cplex.Algorithm.None);
                   break;
         case 'n': cplex.SetParam(Cplex.IntParam.RootAlg,
                                  Cplex.Algorithm.Network);
                   break;
         default:  Usage();
                   return;
         }

         cplex.ImportModel(args[0]);


         if ( cplex.Solve() ) {
            System.Console.WriteLine("Solution status = " + cplex.GetStatus());
            System.Console.WriteLine("Solution value  = " + cplex.ObjValue);

            IEnumerator matrixEnum = cplex.GetLPMatrixEnumerator();
            matrixEnum.MoveNext();

            ILPMatrix lp = (ILPMatrix)matrixEnum.Current;

            double[] x = cplex.GetValues(lp);
            for (int j = 0; j < x.Length; ++j) {
               System.Console.WriteLine("Variable " + j + ": Value = " + x[j]);
            }
         }
         cplex.End();
      }
      catch (ILOG.Concert.Exception e) {
         System.Console.WriteLine("Concert exception caught: " + e);
      }
   }
}

File: Progs/PhD/src/ILP/examples/src/vb/QPex2.vb
' ---------------------------------------------------------------------------
' File: QPex2.vb
' Version 12.2  
' ---------------------------------------------------------------------------
' Licensed Materials - Property of IBM
' 5725-A06 5725-A29 5724-Y48 5724-Y49 5724-Y54 5724-Y55
' Copyright IBM Corporation 2001, 2010. All Rights Reserved.
' 
' US Government Users Restricted Rights - Use, duplication or
' disclosure restricted by GSA ADP Schedule Contract with
' IBM Corp.
' ---------------------------------------------------------------------------
'
' QPex2.vb - Reading in and optimizing a QP problem
'
' To run this example, command line arguments are required.
' i.e.,   QPex2  filename method
' where 
'     filename is the name of the file, with .mps, .lp, or .sav extension
'     method   is the optimization method
'                 o          default
'                 p          primal simplex
'                 d          dual   simplex
'                 b          barrier without crossover
'                 n          network with dual simplex cleanup
'
' Example:
'     QPex2  qpex.lp o
'
Imports ILOG.Concert
Imports ILOG.CPLEX
Imports System.Collections


Public Class QPex2
   
   Friend Shared Sub Usage()
      System.Console.WriteLine("usage:  QPex2 <filename> <method>")
      System.Console.WriteLine("          o       default")
      System.Console.WriteLine("          p       primal simplex")
      System.Console.WriteLine("          d       dual   simplex")
      System.Console.WriteLine("          b       barrier without crossover")
      System.Console.WriteLine("          n       network with dual simplex cleanup")
   End Sub 'Usage
   
   Public Overloads Shared Sub Main(ByVal args() As String)
      If args.Length <> 2 Then
         Usage()
         Return
      End If
      Try
         Dim cplex As New Cplex()

         ' Evaluate command line option and set optimization method accordingly.
         Select Case args(1).ToCharArray()(0)
            Case "o"c
               cplex.SetParam(cplex.IntParam.RootAlg, cplex.Algorithm.Auto)
            Case "p"c
               cplex.SetParam(cplex.IntParam.RootAlg, cplex.Algorithm.Primal)
            Case "d"c
               cplex.SetParam(cplex.IntParam.RootAlg, cplex.Algorithm.Dual)
            Case "b"c
               cplex.SetParam(cplex.IntParam.RootAlg, cplex.Algorithm.Barrier)
               cplex.SetParam(cplex.IntParam.BarCrossAlg, cplex.Algorithm.None)
            Case "n"c
               cplex.SetParam(cplex.IntParam.RootAlg, cplex.Algorithm.Network)
            Case Else
               Usage()
               Return
         End Select

         cplex.ImportModel(args(0))

         If cplex.Solve() Then
            System.Console.WriteLine(("Solution status = " + cplex.GetStatus().ToString))
            System.Console.WriteLine(("Solution value  = " & cplex.ObjValue))

            Dim matrixEnum As IEnumerator = cplex.GetLPMatrixEnumerator()
            matrixEnum.MoveNext()

            Dim lp As ILPMatrix = CType(matrixEnum.Current, ILPMatrix)

            Dim x As Double() = cplex.GetValues(lp)
            Dim j As Integer

            For j = 0 To x.Length - 1
               System.Console.WriteLine(("Variable " & j & ": Value = " & x(j)))
            Next j
         End If
         cplex.End()
      Catch e As ILOG.Concert.Exception
         System.Console.WriteLine(("Concert exception caught: " + e.ToString))
      End Try
   End Sub 'Main
End Class 'QPex2
